olive/evaluator/metric.py

```diff
@@ -47,11 +50,12 @@ class LatencySubType(str, Enum):
 class Metric(ConfigBase):
     name: str
     type: MetricType
-    sub_type: Union[AccuracySubType, LatencySubType] = None
+    sub_type: Union[List[Union[str, AccuracySubType, LatencySubType]], AccuracySubType, LatencySubType] = None
```

Comment:
Do we consider multiple metrics with same type but different sub_types as a same `Metric` object? Should they be different metrics? I agree we could avoid duplicated calculation for same sub type of a same type metric, but have concerns here if they are considered as a same `Metric`.

Reply:
Here is the motivation for this PR:
1. In model training case, the framework(torch/tf) will minimize the model loss(!= metrics). When the model loss is expected, the final metrics are measured. We can treat `sub_type_for_rank`(_maybe no need_) as model loss here, and `sub_type` list as final metrics shown in dashboard.
2. We should ensure that each `Metric` object is optimization direction even it has many sub_type to measure.

Reply:
Also, customer' request is more like:
they know which metrics they want to optimized but at the same time they also want to double check other metrics, just a quick look. Take `latency` as examples, the percentile numbers will help users to check avg numbers is stat-sig or not.

Reply:
I think the idea of only using one sub type for ranking is limiting and only serves the single customer request you describe. 
The point about having the same optimization direction doesn't take into account that the relationship between the subtypes is not always linear. For example, most of the time when precision increases, recall decreases. 

I think the customer request/use case that you describe could be fulfilled if we allows for a value like `-1` subtype priority rank for metric subtypes we want to evaluate but not use for optimization or ranking. 
This meets the above use case while also being more useful. For example, we might want to optimize both precision and recall at the same time. 

Custom metrics can also return multiple metric values that can be ignored or used for optimization and ranking.

Reply:
Yeah I agree, metrics like `precision`, `f1`, `recall`, `auc` they are not quite same. If I'm a customer, I do have a use case in my mind that I would want a separate priority rank for each metric to my preference. `sub_type_for_rank` may not be sufficient for this use case.

Reply:
Yeah, indeed, there always be the case about joint optimization like to optimize `precision` and `recall` at the same time. 

I were thinking that in **current main branch**, we keep the `metric` concept as a single optimization goal. If we want to want to make joint metrics optimization, we have to define different metrics like:
`['metric1': {...}, 'metric2': {...}]`. Then olive will run search strategy to jointly optimize `metric1` and `metric2`.

Reply:
Then for your suggestions, the thing would be like following, 
`
[
  'metric1': {
    'sub_types': {
      'precision': {'rank': 1}
      'recall': {'rank': 2}
      f1': {} // <---------- default rank -1 not for optimization.
    }
  }, 
  'metric2': {
   'sub_types': {
     'latency_avg': {'rank': 3}
     'throughput': {'rank': 4}
   }
 }, 
]
`
In this way, we will transfer the `priority_rank` from metric level to `sub_type` level.

If so, really sounds good than single `sub_type_for_rank`, hahahah. 🤣  I were thinking to keep the original `Metric` definition then stuck in `sub_type_for_rank`.

Reply:
Updated.